### ISSUE TITLE
sendmsg.c incorrectly declares data_len to be an int instead of a Py_ssize_t

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5867]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#5867|
|Type|defect|
|Created|2012-08-14 22:33:53Z|
|Branch|https://github.com/twisted/twisted/tree/Py_ssize_t-5867|

[As reported in a few particularly inscrutable bugs in Calendar Server](http://trac.calendarserver.org/ticket/448), there's an incorrectly declared variable in `sendmsg.c` which can lead to some troubles.  I will be bringing over [the patch from Calendar Server's fix.](http://trac.calendarserver.org/changeset/9700)

<details><summary>Searchable metadata</summary>

```
trac-id__5867 5867
type__defect defect
reporter__glyph glyph
priority__highest highest
milestone__ 
branch__branches_Py_ssize_t_5867 branches/Py_ssize_t-5867
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1344983633000000 1344983633000000
changetime__1345532360000000 1345532360000000
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

(In [35206]) Branching to 'Py_ssize_t-5867'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph removed owner|
|-|-|

Due to the vagaries of undefined behavior in C, I believe the emitted _assembler_ is actually correct even though the _C source_ is wrong, on most compiler versions.  So, there are, sadly, no tests for this change and I cannot think of any that I could write.

Is it even _possible_ to unit-test the type-correctness of C code?  Science may never know.

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/Py_ssize_t-5867).

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_3"></a>|@rodrigc commented|
|-|-|

I am new to Twisted and am participating in the Python Twisted Sprint in San Francisco, August 19, 2012.

I am not sure if your fix is correct.

On Linux, the type of the len field for cmsg_hdr is socklen_t:
   http://www.kernel.org/doc/man-pages/online/pages/man3/cmsg.3.html

   and the type of socklen_t, more or less maps to unsigned int:
   http://lxr.free-electrons.com/source/include/asm-generic/posix_types.h#L71

On FreeBSD, the type of the len field for cmsg_hdr is also socklen_t:
   http://fxr.watson.org/fxr/source/sys/socket.h?v=FREEBSD9#L481

   and the type of socklen_t is uint32_t:
   http://fxr.watson.org/fxr/source/sys/_types.h?v=FREEBSD9;im=bigexcerpts#L59


On MacOS X, type type of the len field in cmsg_hdr is u_int:
https://developer.apple.com/library/mac/#documentation/Darwin/Reference/ManPages/man2/recv.2.html



So basically, unless you want to have nasty
platform-specific #ifdef code, I recommend
that you go with the length parameter being a 32-bit unsigned quantity.


Based on this, I recommend changing the code to:

    PY_UINT32_T data_len;


Py_ssize_t is a signed quantity, so you don't want to use that.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc commented|
|-|-|

OK, on second look at this code, I believe your fix is correct.
PyArg_ParseTuple needs to have a Py_ssize_t, as you have done.

This code has already been imported into trunk, via this commit:

------------------------------------------------------------------------
r34070 | exarkun | 2012-04-04 12:41:07 -0700 (Wed, 04 Apr 2012) | 1 line

Import the sendmsg and recvmsg wrappers from Apple CalendarServer, unmodified so far.
------------------------------------------------------------------------

So I recommend that this bug be reassigned to glyph, and marked for closure.

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|PenguinOfDoom set owner to @glyph|
|-|-|

This looks correct.

With PY_SSIZE_T_CLEAN defined, PyArg_ParseTuple's "#" format stores a value of size Py_ssize_t. If the pointer actually holds the address of an int on the stack, the next 4 bytes will be overwritten, resulting in undefined behavior.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set status to `closed`|
|-|-|

(In [35312]) Merge Py_ssize_t-5867

Fix the sizing of variables whose addresses are passed to `PyArg_ParseTuple` in `sendmsg.c`.

Author: glyph

Reviewer: PenguinOfDoom

Fixes: [#5867](https://github.com/twisted/twisted/issues/5867)

`int` and `Py_ssize_t` are not the same width, unfortunately.  The documentation on this is misleading, but any format character that includes a `#` is a `Py_ssize_t`.
